### PR TITLE
refactor: allow training on any supported data type

### PR DESCRIPTION
**Pull request by aditya-neuraco**
### Features

 - Allow training on any data types, removes image and state coupling. 

### Items

 - [NRL-421](https://neuracore.atlassian.net/browse/NCRL-421)

### CODE REVIEW

File: neuracore/ml/algorithms/cnnmlp/cnnmlp.py [hunk 1]
```diff
@@ -695,10 +706,11 @@ def training_step(self, batch: BatchedTrainingSamples) -> BatchedTrainingOutputs
             batch_size=batch.batch_size,
         )
 
-        if set(batch.outputs.keys()) != set(self.output_data_types):
+        if not set(self.output_data_types).issubset(set(batch.outputs.keys())):
```
Comment (kwangneuraco) [inline neuracore/ml/algorithms/cnnmlp/cnnmlp.py]: Shouldn't it be `if not set(batch.outputs.keys()) .issubset(set(self.output_data_types)):`

File: neuracore/ml/algorithms/groot/groot.py [hunk 2]
```diff
@@ -611,21 +608,48 @@ def _combine_proprio(self, batch: BatchedInferenceInputs) -> torch.Tensor:
...
+            return self._pad_state(
+                torch.zeros(
+                    (len(batch), self.actual_state_dim),
+                    dtype=torch.float32,
+                    device=self.device,
+                )
+            )
 
         all_proprio = torch.cat(proprio_list, dim=-1)  # (B, actual_state_dim)
+        if self.proprio_normalizer is None:
```
Comment (kwangneuraco) [inline neuracore/ml/algorithms/groot/groot.py:648]: I find it confusing, in line 357 it can be none

File: neuracore/ml/algorithms/pi0/pi0.py [hunk 3]
```diff
@@ -419,6 +430,8 @@ def _prepare_rgb_images(
             Tuple of (images, masks) where images is a list of tensors
             [B, C, H, W] per camera and masks is a list of [B] tensors.
         """
+        if DataType.RGB_IMAGES not in self.input_data_types:
```
Comment (kwangneuraco) [inline neuracore/ml/algorithms/pi0/pi0.py:437]: If rgb_images are not in the inputs, you can simulate a single all black image otherwise pi0 will raise an error

File: neuracore/ml/algorithms/pi05/pi05.py [hunk 4]
```diff
@@ -453,6 +471,8 @@ def _prepare_rgb_images(
             Tuple of (images, masks) where images is a list of tensors
             [B, C, H, W] per camera and masks is a list of [B] tensors.
         """
+        if DataType.RGB_IMAGES not in self.input_data_types:
```
Comment (kwangneuraco) [inline neuracore/ml/algorithms/pi05/pi05.py:469]: same as pi0

File: neuracore/ml/algorithms/groot/groot.py (hunk 2 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/groot/groot.py:648]: Yeah, made it a bit clearer now.

File: neuracore/ml/datasets/pytorch_dummy_dataset.py [hunk 5]
```diff
@@ -18,6 +18,7 @@
     DataType,
     NCDataStats,
 )
+from neuracore_types.nc_data import NCDataStatsUnion
```
Comment (kwangneuraco) [inline neuracore/ml/datasets/pytorch_dummy_dataset.py]: Does this related to this PR? Or you need a separate PR for this

File: neuracore/ml/algorithms/act/act.py [hunk 6]
```diff
@@ -67,6 +67,7 @@ class ACT(NeuracoreModel):
     def __init__(
         self,
         model_init_description: ModelInitDescription,
+        *,
```
Comment (kwangneuraco) [inline neuracore/ml/algorithms/act/act.py]: why this is needed?

File: neuracore/ml/algorithms/pi0/pi0.py [hunk 7]
```diff
@@ -89,6 +90,11 @@ def __init__(
         lr_scheduler_decay_lr: float = 2.5e-6,
         finetune_action_expert_only: bool = False,
         freeze_language_model_only: bool = False,
+        vision_num_hidden_layers: int | None = None,
+        vision_hidden_size: int | None = None,
+        vision_intermediate_size: int | None = None,
+        vision_num_attention_heads: int | None = None,
+        vision_patch_size: int | None = None,
```
Comment (kwangneuraco) [inline neuracore/ml/algorithms/pi0/pi0.py]: I didn't put this here because for users they normally don't care about it, adding them to the hyperparameter will confuse them

File: neuracore/ml/algorithms/pi0/pi0.py [hunk 8]
```diff
@@ -148,6 +164,22 @@ def __init__(
         self.finetune_action_expert_only = finetune_action_expert_only
         self.freeze_language_model_only = freeze_language_model_only
 
+        vision_overrides = [
```
Comment (kwangneuraco) [inline neuracore/ml/algorithms/pi0/pi0.py]: again i don't think this should happen here, if any pro-expert wants to modify it, they can create their own algorithm

File: neuracore/ml/algorithms/pi0/pi0.py [hunk 9]
```diff
@@ -419,8 +467,20 @@ def _prepare_rgb_images(
...
             [B, C, H, W] per camera and masks is a list of [B] tensors.
         """
+        if DataType.RGB_IMAGES not in self.input_data_types:
+            return [], []
         if DataType.RGB_IMAGES not in batch.inputs:
-            raise ValueError("RGB images are required but not provided")
+            batch_size = len(batch)
+            black = torch.full(
+                (batch_size, 3, *IMAGE_RESIZE_SHAPE),
+                -1.0,
```
Comment (kwangneuraco) [inline neuracore/ml/algorithms/pi0/pi0.py:443]: check the normalised camera value range is [-1, 1] or [0, 1]

File: neuracore/ml/algorithms/pi0/pi0.py [hunk 10]
```diff
@@ -596,6 +656,10 @@ def training_step(self, batch: BatchedTrainingSamples) -> BatchedTrainingOutputs
         Returns:
             BatchedTrainingOutputs: Training outputs with losses and metrics
         """
+        self.model.train()
```
Comment (kwangneuraco) [inline neuracore/ml/algorithms/pi0/pi0.py]: why you have this?

Comment (kwangneuraco) [reply neuracore/ml/algorithms/pi0/pi0.py]: This is not related to the topic of this PR

File: neuracore/ml/algorithms/pi0/pi0.py [hunk 11]
```diff
@@ -419,8 +467,20 @@ def _prepare_rgb_images(
             Tuple of (images, masks) where images is a list of tensors
             [B, C, H, W] per camera and masks is a list of [B] tensors.
         """
```
Comment (kwangneuraco) [inline neuracore/ml/algorithms/pi0/pi0.py:432]: Give comments on why we have self.input_data_types and batch.inputs separately. One for training and another for inference.

File: neuracore/ml/algorithms/pi0/pi0.py [hunk 12]
```diff
@@ -596,6 +656,10 @@ def training_step(self, batch: BatchedTrainingSamples) -> BatchedTrainingOutputs
         Returns:
             BatchedTrainingOutputs: Training outputs with losses and metrics
         """
+        self.model.train()
+        if self.gradient_checkpointing:
```
Comment (kwangneuraco) [inline neuracore/ml/algorithms/pi0/pi0.py]: Also this is enabled inside the class, this is double code

File: neuracore/ml/algorithms/act/act.py [hunk 13]
```diff
@@ -390,38 +393,46 @@ def _encode_latent(
         along with proprioceptive state into latent distribution parameters.
 
         Args:
-            state: Proprioceptive state features
+            state: Proprioceptive state features (may be None for visual-only)
```
Comment (mark-neuracore) [inline neuracore/ml/algorithms/act/act.py]: Add None to the type to indicate that in the function signaure

File: neuracore/ml/algorithms/pi0/pi0.py [hunk 14]
```diff
@@ -606,16 +670,26 @@ def training_step(self, batch: BatchedTrainingSamples) -> BatchedTrainingOutputs
...
-                f" Expected {self.output_data_types}, got {list(batch.outputs.keys())}"
-            )
-
-        # Concatenate all output actions
+        # Concatenate all output actions; zero-fill types absent from this batch
         action_targets = []
+        action_dim_mask = torch.ones(self.action_dim, device=self.device)
         for data_type in self.ordered_output_data_types:
-            if data_type in [DataType.JOINT_TARGET_POSITIONS, DataType.JOINT_POSITIONS]:
+            start, end = self.output_dims[data_type]
```
Comment (kwangneuraco) [inline neuracore/ml/algorithms/pi0/pi0.py:642]: We don't need it. It is training, and the old data already does the masking, what is the reason of this extra code?

File: neuracore/ml/algorithms/act/act.py (hunk 6 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/act/act.py]: There are a large number of arguments. This is to force the user to call the class by keyword arguments. It's good code hygeine practice. Large number of positional arguments are brittle to refactors.

File: neuracore/ml/algorithms/pi0/pi0.py (hunk 7 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/pi0/pi0.py]: It's completely optional.  I added this to make the vision encoder configurable. Ideally, all neuracore algorithms are meant to be general, configurable implementations of seminal algorithms.

This is passed so that tests can bu run with a tiny vision encoder.

File: neuracore/ml/algorithms/pi0/pi0.py (hunk 8 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/pi0/pi0.py]: We want our algorithms to be as configurable as possible. That's why I added it. This also speeds up testing.

File: neuracore/ml/algorithms/pi0/modules.py [hunk 15]
```diff
@@ -71,6 +71,11 @@ def __init__(self, config: PI0Config):
             action_expert_config,
             use_adarms=config.use_adarms,
             precision=config.dtype,
+            vision_num_hidden_layers=config.vision_num_hidden_layers,
```
Comment (kwangneuraco) [inline neuracore/ml/algorithms/pi0/modules.py]: Unless you want to change the sigLip model, No one is going to use it, don't add this

File: neuracore/ml/algorithms/pi0/utils.py [hunk 16]
```diff
@@ -71,6 +71,11 @@ class PI0Config:
     input_features: dict = field(default_factory=dict)
     output_features: dict = field(default_factory=dict)
     image_features: list[str] = field(default_factory=list)
+    vision_num_hidden_layers: int | None = None
```
Comment (kwangneuraco) [inline neuracore/ml/algorithms/pi0/utils.py]: Same

File: neuracore/ml/algorithms/pi0/gemma_pytorch.py [hunk 17]
```diff
@@ -219,6 +219,12 @@ def __init__(
         action_expert_config: GemmaConfig,
         use_adarms: tuple[bool, bool] | None = None,
         precision: Literal["bfloat16", "float32"] = "bfloat16",
+        *,
```
Comment (kwangneuraco) [inline neuracore/ml/algorithms/pi0/gemma_pytorch.py]: same here

File: neuracore/ml/algorithms/pi0/pi0.py (hunk 12 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/pi0/pi0.py]: The model enables, disables, multiple times (4 times in total). This is just to make everything consistent.

File: neuracore/ml/algorithms/pi0/pi0.py (hunk 10 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/pi0/pi0.py]: Ok.

File: neuracore/ml/algorithms/act/act.py (hunk 6 above)
Comment (mark-neuracore) [reply neuracore/ml/algorithms/act/act.py]: This is to enforce passing with keywords. It is important but IMO out of the scope of this PR.

File: neuracore/ml/algorithms/act/act.py (hunk 13 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/act/act.py]: ok

File: neuracore/ml/algorithms/pi0/modules.py (hunk 15 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/pi0/modules.py]: These changes are mostly to speedup tests, similar to groot.

File: neuracore/ml/algorithms/pi05/pi05.py [hunk 18]
```diff
@@ -93,6 +94,11 @@ def __init__(
         finetune_action_expert_only: bool = False,
         finetune_vision_encoder_and_action_expert: bool = False,
         discrete_state_input: bool = True,
+        vision_num_hidden_layers: int | None = None,
+        vision_hidden_size: int | None = None,
+        vision_intermediate_size: int | None = None,
+        vision_num_attention_heads: int | None = None,
+        vision_patch_size: int | None = None,
```
Comment (kwangneuraco) [inline neuracore/ml/algorithms/pi05/pi05.py]: I think we would better move this to another PR for later disucssion

File: neuracore/ml/datasets/pytorch_dummy_dataset.py (hunk 5 above)
Comment (aditya-neuraco) [reply neuracore/ml/datasets/pytorch_dummy_dataset.py]: Moved it to a separate branch.

File: neuracore/ml/algorithms/act/act.py (hunk 6 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/act/act.py]: At second thought, I agree we should move it to a separate PR.

File: neuracore/ml/algorithms/pi0/pi0.py [hunk 19]
```diff
@@ -638,8 +712,11 @@ def training_step(self, batch: BatchedTrainingSamples) -> BatchedTrainingOutputs
         mse_losses = self.model.forward(
             images, image_masks, lang_tokens, lang_masks, proprios, target_actions
         )
-        # Mask to the real action dims
-        loss = mse_losses[:, :, : self.action_dim].mean()
+        # Mask to real action dims, excluding types absent from this batch
```
Comment (kwangneuraco) [inline neuracore/ml/algorithms/pi0/pi0.py]: Maybe we don't need the mask of actions because the original code did it

File: neuracore/ml/algorithms/pi0/pi0.py (hunk 7 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/pi0/pi0.py]: Moved to a separate branch.

File: neuracore/ml/algorithms/pi0/pi0.py (hunk 9 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/pi0/pi0.py:443]: It's `[-1, 1]`

File: neuracore/ml/algorithms/pi0/pi0.py (hunk 8 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/pi0/pi0.py]: Removed from this PR. See: https://github.com/NeuracoreAI/neuracore/pull/629

File: neuracore/ml/algorithms/pi0/pi0.py (hunk 12 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/pi0/pi0.py]: Discusses offline.

File: neuracore/ml/algorithms/pi0/modules.py (hunk 15 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/pi0/modules.py]: Moved elsewhere, See: https://github.com/NeuracoreAI/neuracore/pull/629

File: neuracore/ml/algorithms/pi0/gemma_pytorch.py (hunk 17 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/pi0/gemma_pytorch.py]: Moved to a different branch.

File: neuracore/ml/algorithms/pi05/pi05.py (hunk 18 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/pi05/pi05.py]: Moved.

File: neuracore/ml/algorithms/pi0/utils.py (hunk 16 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/pi0/utils.py]: Moved.

**Review by kwangneuraco — APPROVED** [review]
LGTM now

**Review by kwangneuraco — APPROVED** [review]
LGTM now

File: neuracore/ml/algorithms/pi0/pi0.py (hunk 19 above)
Comment (aditya-neuraco) [reply neuracore/ml/algorithms/pi0/pi0.py]: We need it. Discussed offline why.
